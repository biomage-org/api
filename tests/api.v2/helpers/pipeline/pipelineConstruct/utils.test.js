const _ = require('lodash');
const AWSMock = require('aws-sdk-mock');
const AWS = require('../../../../../src/utils/requireAWS');

const { createNewStateMachine } = require('../../../../../src/api.v2/helpers/pipeline/pipelineConstruct/utils');

const mockCluster = {
  cluster: {
    name: 'biomage-test',
    endpoint: 'https://test-endpoint.me/fgh',
    certificateAuthority: {
      data: 'AAAAAAAAAAA',
    },
  },
};

describe('createNewStateMachine', () => {
  afterEach(() => {
    AWSMock.restore('EKS');
    AWSMock.restore('StepFunctions');
  });

  it('State machine is updated instead of created if it already exists.', async () => {
    AWSMock.setSDKInstance(AWS);

    const describeClusterSpy = jest.fn((x) => x);
    AWSMock.mock('EKS', 'describeCluster', (params, callback) => {
      describeClusterSpy(params);
      callback(null, mockCluster);
    });

    const createStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'createStateMachine', (params, callback) => {
      createStateMachineSpy(params);
      callback({ code: 'StateMachineAlreadyExists' }, null);
    });

    const updateStateMachineSpy = jest.fn((stateMachineObject) => _.omit(stateMachineObject, 'definition'));
    AWSMock.mock('StepFunctions', 'updateStateMachine', (params, callback) => {
      updateStateMachineSpy(params);
      callback(null, { stateMachineArn: 'test-machine' });
    });

    const context = { experimentId: 'mockExperimentId', roleArn: 'mockRoleArn', accountId: 'mockAccountId' };
    const stateMachineArn = 'mockStateMachineArn';

    await createNewStateMachine(context, stateMachineArn, 'SomePipelineName');

    expect(describeClusterSpy).toMatchSnapshot();
    expect(createStateMachineSpy.mock.results).toMatchSnapshot();

    expect(updateStateMachineSpy).toHaveBeenCalled();
    expect(updateStateMachineSpy.mock.results).toMatchSnapshot();
  });
});
