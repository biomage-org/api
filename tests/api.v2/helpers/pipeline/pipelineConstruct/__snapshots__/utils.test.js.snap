// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createNewStateMachine State machine is updated instead of created if it already exists. 1`] = `[MockFunction]`;

exports[`createNewStateMachine State machine is updated instead of created if it already exists. 2`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "loggingConfiguration": Object {
        "level": "OFF",
      },
      "name": "biomage-SomePipelineName-test-6b71d11936e7d590b1e0c8c1f01bd6abef203bab",
      "roleArn": "mockRoleArn",
      "tags": Array [
        Object {
          "key": "experimentId",
          "value": "mockExperimentId",
        },
        Object {
          "key": "clusterEnv",
          "value": "test",
        },
        Object {
          "key": "sandboxId",
          "value": "default",
        },
      ],
      "type": "STANDARD",
    },
  },
]
`;

exports[`createNewStateMachine State machine is updated instead of created if it already exists. 3`] = `
Array [
  Object {
    "type": "return",
    "value": Object {
      "roleArn": "mockRoleArn",
      "stateMachineArn": "arn:aws:states:eu-west-1:mockAccountId:stateMachine:biomage-SomePipelineName-test-6b71d11936e7d590b1e0c8c1f01bd6abef203bab",
    },
  },
]
`;
